src/v/transform/transfer_queue.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "base/seastarx.h"
+#include "ssx/semaphore.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/chunked_fifo.hh>
+#include <seastar/core/condition-variable.hh>
+
+#include <algorithm>
+#include <optional>
+
+namespace transform {
+
+template<typename T>
+concept MemoryMeasurable = requires(const T v) {
+    { v.memory_usage() } -> std::convertible_to<size_t>;
+};
+
+constexpr size_t default_items_per_chunk = 128;
+
+template<MemoryMeasurable T, size_t items_per_chunk = default_items_per_chunk>
+class transfer_queue {
+public:
+    explicit transfer_queue(size_t max_memory_usage)
+      : _max_memory(max_memory_usage)
+      , _memory_sem(max_memory_usage, "transform::transfer_queue") {}
+
+    ss::future<> push(T entry) {
+        // We want to be able to always insert at least one item, so cap memory
+        // usage by _max_memory so we don't overload our semaphore and so we
+        // can't get stuck.
+        size_t mem = std::min(entry.memory_usage(), _max_memory);
```

Comment:
Yeah we have to make progress, so eating all the mutex is important. Overall the memory limits are still in place because I think somewhere else needs to have already accounted for the memory here...

Reply:
>somewhere else needs to have already accounted for the memory here

You mean when the event was constructed? That seems like an easy precondition to break. This is sort of confusing.

Reply:
> You mean when the event was constructed? That seems like an easy precondition to break. This is sort of confusing.

Yeah the transform subsystem needs to have a limit on the total memory it uses for what is available - it does not right now, we have a limit defined, but don't really respect it. At the end of the day batches are variable sized, and it feels odd to have a fixed buffer size when each batch could have wildly different actual memory costs.

However we can't not make progress because we get a batch that is over our limit, we just need to go slower. This is sort of how our log reader uses the max_bytes limit FWIW, it's a soft limit and you'll always make progress over adhering to the limit.

Let's continue this in the parent conversation.

Reply:
Ok, yeah that makes good sense 👍

Reply:
Let me expand more on this in the comments for this class